### PR TITLE
Avoid cbuffer copy for special pattern.

### PR DESCRIPTION
**Pull request by python3kgae**
float4 cb[16];
float v[64] = cb;

### CODE REVIEW

File: tools/clang/lib/CodeGen/CGHLSLMS.cpp [hunk 1]
```diff
@@ -5371,6 +5385,29 @@ void CGMSHLSLRuntime::EmitHLSLFlatConversionAggregateCopy(CodeGenFunction &CGF,
...
       CGF.Builder.CreateMemCpy(DestPtr, SrcPtr, std::min(sizeSrc, sizeDest), 1);
       return;
+    } else if (GlobalVariable *SrcGV = dyn_cast<GlobalVariable>(SrcPtr)) {
+      if (m_ConstVarAnnotationMap.count(SrcGV) &&
+          IsVec4ArrayToScalarArrayForMemcpy(SrcPtrTy, DestPtrTy)) {
+        unsigned sizeSrc = TheModule.getDataLayout().getTypeAllocSize(SrcPtrTy);
+        unsigned sizeDest =
+            TheModule.getDataLayout().getTypeAllocSize(DestPtrTy);
+        if (sizeSrc == sizeDest) {
+          CGF.Builder.CreateMemCpy(DestPtr, SrcPtr, std::min(sizeSrc, sizeDest),
```
Comment (tex3d) [inline tools/clang/lib/CodeGen/CGHLSLMS.cpp:5395]: Why do `std::min(sizeSrc, sizeDest)` when `sizeSrc == sizeDest`?  Same for the next case.

File: tools/clang/lib/CodeGen/CGHLSLMS.cpp [hunk 2]
```diff
@@ -5371,6 +5385,29 @@ void CGMSHLSLRuntime::EmitHLSLFlatConversionAggregateCopy(CodeGenFunction &CGF,
...
+                                   1);
+          return;
+        }
+      }
+    }
+  } else if (GlobalVariable *SrcGV = dyn_cast<GlobalVariable>(SrcPtr)) {
```
Comment (tex3d) [inline tools/clang/lib/CodeGen/CGHLSLMS.cpp:5401]: So this case is copy to local?  Test?

File: tools/clang/test/HLSLFileCheck/hlsl/objects/CbufferLegacy/cb_array_mutate.hlsl [hunk 3]
```diff
@@ -0,0 +1,19 @@
...
+// Make sure only 1 cb load.
+// CHECK:call %dx.types.CBufRet.i32 @dx.op.cbufferLoadLegacy.i32
+// CHECK-NOT:call %dx.types.CBufRet
+
+cbuffer Pack
+{
+    int4 __packed[16];
+};
+
+static int arrayReallyWant[64] = (int[64])__packed;
```
Comment (tex3d) [inline tools/clang/test/HLSLFileCheck/hlsl/objects/CbufferLegacy/cb_array_mutate.hlsl:15]: What happens with groupshared?  Local?  How does this handle modification of dest array? Should this require const on the dest array?

File: tools/clang/lib/CodeGen/CGHLSLMS.cpp (hunk 2 above)
Comment (python3kgae) [reply tools/clang/lib/CodeGen/CGHLSLMS.cpp]: No. Disabled.

File: tools/clang/test/HLSLFileCheck/hlsl/objects/CbufferLegacy/cb_array_mutate.hlsl (hunk 3 above)
Comment (python3kgae) [reply tools/clang/test/HLSLFileCheck/hlsl/objects/CbufferLegacy/cb_array_mutate.hlsl:15]: Only static for now.

**Review by tex3d — APPROVED** [review]

**Pull request merged**